### ISSUE TITLE
GCStress failure: JIT\HardwareIntrinsics\X86\Sse\Sse_r

### ISSUE THREAD

**Issue by BruceForstall** [dotnet/coreclr#16370]
Failure with `COMPlus_GCStress=4` with x86 Checked with JIT\HardwareIntrinsics\X86\Sse\Sse_r:

```
BEGIN EXECUTION
 "e:\gh\coreclr3\bin\tests\Windows_NT.x86.Checked\Tests\Core_Root\corerun.exe" Sse_r.exe
Running Add.Single test...
Sse.Add<Single>: RunClsVarScenario failed:
    left: (0, 0, 0, 0)
   right: (0.8884231, 0.7757907, 0.8753635, 0.08364759)
  result: (1.688369, 1.052082, 1.488229, 0.1442138)

Sse.Add<Single>: RunLclVarScenario_UnsafeRead failed:
    left: (0.9750369, 0.8985233, 0.3467779, 0.7324353)
   right: (0.887246, 0.08891934, 0.3435928, 0.5924375)
  result: (1.862283, 0.9874426, 0.6903708, 1.324873)

Sse.Add<Single>: RunLclVarScenario_Load failed:
    left: (0.9750369, 0.8985233, 0.3467779, 0.7324353)
   right: (0.887246, 0.08891934, 0.3435928, 0.5924375)
  result: (1.862283, 0.9874426, 0.6903708, 1.324873)

Sse.Add<Single>: RunLclVarScenario_LoadAligned failed:
    left: (0.9750369, 0.8985233, 0.3467779, 0.7324353)
   right: (0.887246, 0.08891934, 0.3435928, 0.5924375)
  result: (1.862283, 0.9874426, 0.6903708, 1.324873)

Sse.Add<Single>: RunLclFldScenario failed:
    left: (0.2974494, 0.9823032, 0.6075894, 0.4913346)
   right: (0.4457132, 0.5806072, 0.989637, 0.6327882)
  result: (0.7431626, 1.56291, 1.597226, 1.124123)

Sse.Add<Single>: RunFldScenario failed:
    left: (0.5911472, 0.9279458, 0.5975241, 0.3108502)
   right: (0.6367142, 0.3006574, 0.2591962, 0.6673818)
  result: (1.227861, 1.228603, 0.8567203, 0.9782319)

Error: One or more scenarios did not complete as expected.
Running AddScalar.Single test...
Sse.AddScalar<Single>: RunClsVarScenario failed:
    left: (0, 0, 0, 0)
   right: (0, 0, 0, 0)
  result: (1.138822, 0.2528097, 0.4958607, 0.5890472)

Sse.AddScalar<Single>: RunLclVarScenario_UnsafeRead failed:
    left: (0.6187453, 0.5449098, 0.8567044, 0.1804269)
   right: (0.3909108, 0.6793364, 0.447796, 0.7770419)
  result: (1.009656, 0.5449098, 0.8567044, 0.1804269)

Sse.AddScalar<Single>: RunLclVarScenario_Load failed:
    left: (0.6187453, 0.5449098, 0.8567044, 0.1804269)
   right: (0.3909108, 0.6793364, 0.447796, 0.7770419)
  result: (1.009656, 0.5449098, 0.8567044, 0.1804269)

Sse.AddScalar<Single>: RunLclVarScenario_LoadAligned failed:
    left: (0.6187453, 0.5449098, 0.8567044, 0.1804269)
   right: (0.3909108, 0.6793364, 0.447796, 0.7770419)
  result: (1.009656, 0.5449098, 0.8567044, 0.1804269)

Sse.AddScalar<Single>: RunLclFldScenario failed:
    left: (0.7431608, 0.3680078, 0.2915384, 0.4571137)
   right: (0.3812023, 0.9455716, 0.04463677, 0.2569035)
  result: (1.124363, 0.3680078, 0.2915384, 0.4571137)

Sse.AddScalar<Single>: RunFldScenario failed:
    left: (0.7790595, 0.6797857, 0.4424345, 0.01052309)
   right: (0.03844457, 0.4287005, 0.9583447, 0.505011)
  result: (0.8175041, 0.6797857, 0.4424345, 0.01052309)

Error: One or more scenarios did not complete as expected.
Running And.Single test...
Running AndNot.Single test...
Sse.AndNot<Single>: RunClsVarScenario failed:
    left: (0.1116372, 0.1851131, 0.3545335, 0.8148552)
   right: (0, 0, 0, 0)
  result: (1.124304E-37, 5.912337E-39, 2.393037E-38, 1.322114E-39)

Sse.AndNot<Single>: RunLclVarScenario_UnsafeRead failed:
    left: (0.01076706, 0.1196732, 0.7541804, 0.1733052)
   right: (0.1081166, 0.2048262, 0.1701058, 0.1039674)
  result: (7.531112E-38, 9.452388E-38, 4.235912E-39, 7.219427E-38)

Sse.AndNot<Single>: RunLclVarScenario_Load failed:
    left: (0.01076706, 0.1196732, 0.7541804, 0.1733052)
   right: (0.1081166, 0.2048262, 0.1701058, 0.1039674)
  result: (7.531112E-38, 9.452388E-38, 4.235912E-39, 7.219427E-38)

Sse.AndNot<Single>: RunLclVarScenario_LoadAligned failed:
    left: (0.01076706, 0.1196732, 0.7541804, 0.1733052)
   right: (0.1081166, 0.2048262, 0.1701058, 0.1039674)
  result: (7.531112E-38, 9.452388E-38, 4.235912E-39, 7.219427E-38)

Sse.AndNot<Single>: RunLclFldScenario failed:
    left: (0.4973004, 0.9443557, 0.3046108, 0.5677655)
   right: (0.3790673, 0.4320269, 0.007059659, 0.4996125)
  result: (3.881597E-43, 1.287513E-38, 4.187699E-38, 2.190425E-38)

Sse.AndNot<Single>: RunFldScenario failed:
    left: (0.6086615, 0.4492103, 0.7428799, 0.04668195)
   right: (0.1993912, 0.1871578, 0.9181196, 0.5265004)
  result: (6.260601E-39, 2.387734E-39, 5.969489E-39, 9.461397E-38)

Error: One or more scenarios did not complete as expected.
Running CompareEqual.Single test...
Running CompareEqualScalar.Single test...
Running CompareEqualOrderedScalar.Single test...
Running CompareEqualUnorderedScalar.Single test...
Running CompareGreaterThan.Single test...
Running CompareGreaterThanScalar.Single test...
Running CompareGreaterThanOrderedScalar.Single test...
Running CompareGreaterThanUnorderedScalar.Single test...
Running CompareGreaterThanOrEqual.Single test...
Running CompareGreaterThanOrEqualScalar.Single test...
Running CompareGreaterThanOrEqualOrderedScalar.Single test...
Running CompareGreaterThanOrEqualUnorderedScalar.Single test...
Running CompareLessThan.Single test...
Running CompareLessThanScalar.Single test...
Running CompareLessThanOrderedScalar.Single test...
Running CompareLessThanUnorderedScalar.Single test...
Running CompareLessThanOrEqual.Single test...
Running CompareLessThanOrEqualScalar.Single test...
Running CompareLessThanOrEqualOrderedScalar.Single test...
Running CompareLessThanOrEqualUnorderedScalar.Single test...
Running CompareNotEqual.Single test...
Running CompareNotEqualScalar.Single test...
Running CompareNotEqualOrderedScalar.Single test...
Running CompareNotEqualUnorderedScalar.Single test...
Running CompareNotGreaterThan.Single test...
Running CompareNotGreaterThanScalar.Single test...
Running CompareNotGreaterThanOrEqual.Single test...
Running CompareNotGreaterThanOrEqualScalar.Single test...
Running CompareNotLessThan.Single test...
Running CompareNotLessThanScalar.Single test...
Running CompareNotLessThanOrEqual.Single test...
Running CompareNotLessThanOrEqualScalar.Single test...
Running CompareOrdered.Single test...
Running CompareOrderedScalar.Single test...
Running CompareUnordered.Single test...
Running CompareUnorderedScalar.Single test...
Running Divide.Single test...
Running DivideScalar.Single test...
Running Max.Single test...
Running MaxScalar.Single test...
Running Min.Single test...
Running MinScalar.Single test...
Running Multiply.Single test...
Running MultiplyScalar.Single test...
Running Or.Single test...
Running Subtract.Single test...
Running SubtractScalar.Single test...
Running Xor.Single test...
Expected: 100
Actual: 0
END EXECUTION - FAILED
FAILED
```

https://ci.dot.net/job/dotnet_coreclr/job/master/job/jitstress/job/x64_checked_windows_nt_gcstress0xc/73/#showFailuresLink
https://ci.dot.net/job/dotnet_coreclr/job/master/job/jitstress/job/x86_checked_windows_nt_gcstress0xc/67/#showFailuresLink

**Comment by BruceForstall** [comment]
@CarolEidt @fiigii @tannergooding

**Comment by fiigii** [comment]
Thanks, I will look into this failure.

**Comment by fiigii** [comment]
I tried on my local machine, this can be fixed by https://github.com/dotnet/coreclr/pull/16365

**Comment by fiigii** [comment]
This failure is not related to https://github.com/dotnet/coreclr/pull/16365. We need another fix.

**Comment by tannergooding** [comment]
@BruceForstall, what is the behavior of the `GCStress=4` mode?

Some of these tests aren't accessing anything outside of a pinned struct in `this` and are actually indicating that the computed result for the given inputs is accurate. For example, I modified one of the failing tests to also print out the expected result values (and their binary representation):
```
Sse.AddScalar<Single>: RunLclVarScenario_UnsafeRead failed:
      left: (0.4583388, 0.1761521, 0.01612143, 0.672764) - (1055566690, 1043620150, 1015288089, 1059863107)
     right: (0.3572484, 0.953283, 0.5085903, 0.8649272) - (1052174658, 1064569434, 1057108729, 1063087070)
    result: (0.8155872, 0.1761521, 0.01612143, 0.672764) - (1062259282, 1043620150, 1015288089, 1059863107)
  expected: (0.8155872, 0.1761521, 0.01612143, 0.672764) - (1062259282, 1043620150, 1015288089, 1059863107)

Sse.AddScalar<Single>: RunLclVarScenario_Load failed:
      left: (0.4583388, 0.1761521, 0.01612143, 0.672764) - (1055566690, 1043620150, 1015288089, 1059863107)
     right: (0.3572484, 0.953283, 0.5085903, 0.8649272) - (1052174658, 1064569434, 1057108729, 1063087070)
    result: (0.8155872, 0.1761521, 0.01612143, 0.672764) - (1062259282, 1043620150, 1015288089, 1059863107)
  expected: (0.8155872, 0.1761521, 0.01612143, 0.672764) - (1062259282, 1043620150, 1015288089, 1059863107)

Sse.AddScalar<Single>: RunLclVarScenario_LoadAligned failed:
      left: (0.4583388, 0.1761521, 0.01612143, 0.672764) - (1055566690, 1043620150, 1015288089, 1059863107)
     right: (0.3572484, 0.953283, 0.5085903, 0.8649272) - (1052174658, 1064569434, 1057108729, 1063087070)
    result: (0.8155872, 0.1761521, 0.01612143, 0.672764) - (1062259282, 1043620150, 1015288089, 1059863107)
  expected: (0.8155872, 0.1761521, 0.01612143, 0.672764) - (1062259282, 1043620150, 1015288089, 1059863107)

Sse.AddScalar<Single>: RunLclFldScenario failed:
      left: (0.1804814, 0.9095355, 0.9705721, 0.5292867) - (1043910686, 1063835473, 1064859497, 1057455957)
     right: (0.6334039, 0.1796429, 0.981744, 0.9808517) - (1059202754, 1043854416, 1065046931, 1065031961)
    result: (0.8138853, 0.9095355, 0.9705721, 0.5292867) - (1062230730, 1063835473, 1064859497, 1057455957)
  expected: (0.8138853, 0.9095355, 0.9705721, 0.5292867) - (1062230730, 1063835473, 1064859497, 1057455957)

Sse.AddScalar<Single>: RunFldScenario failed:
      left: (0.3298608, 0.6303476, 0.09755851, 0.2524249) - (1051255685, 1059151478, 1036504257, 1048657365)
     right: (0.4580787, 0.406973, 0.5403498, 0.9132994) - (1055557964, 1053843140, 1057641565, 1063898622)
    result: (0.7879395, 0.6303476, 0.09755851, 0.2524249) - (1061795432, 1059151478, 1036504257, 1048657365)
  expected: (0.7879395, 0.6303476, 0.09755851, 0.2524249) - (1061795432, 1059151478, 1036504257, 1048657365)
```

The one where it doesn't line up (`RunClsVarScenario`) looks correct if you step through with a debugger, but gets messed up at some point before the validation happens:
```
Sse.AddScalar<Single>: RunClsVarScenario failed:
      left: (0, 0, 0, 0) - (0, 0, 0, 0)
     right: (0.7696341, 0.7332616, 0.6672596, 0.9649948) - (1061488318, 1060878089, 1059770759, 1064765926)
    result: (1.047543, 0.3547822, 0.1568179, 0.28519) - (1065752034, 1052091908, 1042322656, 1049756781)
  expected: (0.7696341, 0, 0, 0) - (1061488318, 0, 0, 0)
```

**Comment by BruceForstall** [comment]
`GCStress=4` performs a GC at every JIT-generated instruction. If the GC info is incorrect, the GC might move an object, but not update a JIT-generated address or register, so the value might change out from under you.

**Comment by tannergooding** [comment]
What is the best way to diagnose these types of issues?

For the `LclVar`, `LclFld`, and `Fld` scenarios, it seems like the error is somewhere in [ValidateResult](https://github.com/dotnet/coreclr/blob/master/tests/src/JIT/HardwareIntrinsics/X86/Sse/AddScalar.Single.cs#L287). At least given that, by that point, all three arrays have been created and have the expected values when printed.

For the `ClsVar` scenario, the error is somewhere between the result being computed and it being written to the array that is used for validating the result (so after https://github.com/dotnet/coreclr/blob/master/tests/src/JIT/HardwareIntrinsics/X86/Sse/AddScalar.Single.cs#L197, but before https://github.com/dotnet/coreclr/blob/master/tests/src/JIT/HardwareIntrinsics/X86/Sse/AddScalar.Single.cs#L271).

**Comment by BruceForstall** [comment]
> What is the best way to diagnose these types of issues?

Sometimes the best way is to read very carefully through the JitDump to ensure that the GC info is correct for each instruction generated. E.g., pay attention to lines like these:
```
gcrReg +[r0]  // register "birth"
gcrReg -[r0]   // register "death"
[08528018] byr var born at [r11-3CH]   // stack var byref birth
```

Sometimes you can use CLR logging with GCinfo dumping, but that can give overwhelming information. E.g.,
```
set COMPlus_LogEnable=1
set COMPlus_LogToFile=1
set COMPlus_LogFile=clr.log
set COMPlus_LogFileAppend=
set COMPlus_LogLevel=a
set COMPlus_LogFacility=80003
```

**Comment by tannergooding** [comment]
Wonder if this is related to https://github.com/dotnet/coreclr/issues/16892

**Comment by tannergooding** [comment]
Fix should be here: https://github.com/dotnet/coreclr/pull/16957

**Comment by AndyAyersMS** [comment]
Can we close this, or does somebody need to verify the fix?

**Comment by fiigii** [comment]
> Can we close this,

Yes, it has been fixed by #16957
